### DISCUSSION TITLE
Interface kernels and interface reaction

### DISCUSSION THREAD

**Discussion by EdSheltonUKAEA** [idaholab/moose#18533] [Q&A General] [answered]
I am trying to understand how interface kernels work and in particular the InterfaceReaction kernel.

I have tried to use the example input syntax given for test/tests/interfacekernels/1d_interface/reaction_1D_steady.i at this link:
https://mooseframework.inl.gov/source/interfacekernels/InterfaceReaction.html

When I tried to run this errors occurred.  The first thing I did was to replace the obsolete(?) "InterfaceDiffusion" kernel with "InterfaceDiffusionBoundaryTerm".
However, trying to run the code still gives error "invalid float syntax for parameter" for the diffusion coefficient variable.
Does this need to be declared as an AuxVariable, or is something else going wrong?

Many thanks in advance for any help and clarifications.
Ed

PS.  I also initially got confused by the fact that in MOOSE source documentation webpage (https://mooseframework.inl.gov/source/index.html) "InterfaceKernels" refers to physical boundaries between different domains (e.g. material surfaces and phase boundaries), but just "Interfaces" uses the word in its computing or coding sense of getting different bits of code to work together.  I wonder if it is possible to make these different uses of the word clearer.

**Comment by GiudGiud** [comment, marked as answer]
Hi Ed

I checked the documentation and a float (double) diffusion coefficient is expected for D. Could you please share that block of your input file here?

On wording, Interfaces is fine. A possible replacement is internal sides but it's wordy. InterfaceKernels clearly does not refer to coupling codes together.

Guillaume

**Reply by EdSheltonUKAEA** [reply]
Here's the blocks of code from the example which make use of the "D" variable.

> [Kernels]
>   [./diff_u]
>     type = MatDiffusion
>     variable = u
>     block = '0'
>     diffusivity = D
>   [../]
>   [./diff_v]
>     type = MatDiffusion
>     variable = v
>     block = '1'
>     diffusivity = D
>   [../]
> []
> 
> [InterfaceKernels]
>   [./interface]
>     type = InterfaceDiffusion
>     variable = u
>     neighbor_var = 'v'
>     boundary = 'primary0_interface'
>     D = D
>     D_neighbor = D
>   [../]
>   [./interface_reaction]
>     type = InterfaceReaction
>     variable = u
>     neighbor_var = 'v'
>     boundary = 'primary0_interface'
>     kf = 1 # Forward reaction rate coefficient
>     kb = 2 # Backward reaction rate coefficient
>   [../]
> []
> 
> [BCs]
>   [./left]
>     type = DirichletBC
>     variable = u
>     boundary = 'left'
>     value = 1
>   [../]
>   [./right]
>     type = DirichletBC
>     variable = v
>     boundary = 'right'
>     value = 0
>   [../]
> []
> 
> [Materials]
>   [./block0]
>     type = GenericConstantMaterial
>     block = '0'
>     prop_names = 'D'
>     prop_values = '4'
>   [../]
>   [./block1]
>     type = GenericConstantMaterial
>     block = '1'
>     prop_names = 'D'
>     prop_values = '2'
>   [../]
> []

**Reply by GiudGiud** [reply]
so D should be a real, not a material property for this kernel. 
So just use '2' and '4' for D_neighbor and D. 

If you want to use a material property you'll have to find another interface kernel or adapt this one. Will be like 3 lines to switch it over.

**Reply by EdSheltonUKAEA** [reply]
OK.  Thanks.
Do you mean I change it to this:

> [./interface]
> type = InterfaceDiffusion
> variable = u
> neighbor_var = 'v'
> boundary = 'primary0_interface'
> D = '4'
> D_neighbor = '2'
> [../]

And I can delete the Materials block completely?

As this is example code taken from the MOOSE website, maybe it's worth changing it there too if that's quick and easy to do.

**Reply by GiudGiud** [reply]
Yes.
For this particular input file you could delete the Materials block if you preferred to just specify the float values in the kernels. 
When you have temperature dependent material properties you will need to use material properties

Where did you find this particular example?
I agree it s worth changing.

**Reply by EdSheltonUKAEA** [reply]
And a new error: 
missing required parameter 'Mesh/interface/master_block' 
        Doc String: "The master set of blocks for which to draw a sideset between"

It's on the website here:
https://mooseframework.inl.gov/source/interfacekernels/InterfaceReaction.html

**Reply by GiudGiud** [reply]
ok so `InterfaceDiffusion` is not obsolete. It's a test object. And it uses material properties for D and D_neighbor. 

If you would rather use that, then you need to run moose with --allow-test-objects

**Reply by GiudGiud** [reply]
this is a separate error in a different block. There should not be a parameter called master_block.. we renamed them to primary block.
Could you pls paste that block here? I m curious

**Reply by GiudGiud** [reply]
Created an issue #18534 to fix that up later

**Reply by EdSheltonUKAEA** [reply]
Thanks for the tip about test objects.  Is there a link to more information about these?  I can't see them specifically mentioned in either the A-Z index or source documentation page.

Below is the mesh block apparently giving the master block error.  As for the other code, I've just copied and pasted it from the example file on the website.

> [Mesh]
>   [gen]
>     type = GeneratedMeshGenerator
>     dim = 1
>     nx = 10
>     xmax = 2
>   []
>   [./subdomain1]
>     input = gen
>     type = SubdomainBoundingBoxGenerator
>     bottom_left = '1.0 0 0'
>     block_id = 1
>     top_right = '2.0 1.0 0'
>   [../]
>   [./interface]
>     type = SideSetsBetweenSubdomainsGenerator
>     input = 'subdomain1'
>     primary_block = '0'
>     paired_block = '1'
>     new_boundary = 'primary0_interface'
>   [../]
> []

**Reply by GiudGiud** [reply]
There is not. They are strictly meant for testing the framework. You can find them in moose/test/src/ though and reading the source code is usually very easy for the small simple objects

How old is your moose version? 
The master_block parameter is retrieved if the primary_block is not provided to SideSetsBetweenSubdomainsGenerator. It seems fine here though.

**Comment by EdSheltonUKAEA** [comment]
Also, I'm not really a computer scientist or coder, so I'm not sure where I'm meant to put that line "--allow-test-objects".  I tried appending it to the the command line in Ubuntu, but that didn't seem to make any difference.  Should it go in one of the MOOSE input file kernels, or one of the files on my PC actually running MOOSE?  Thanks again!

**Reply by GiudGiud** [reply]
This goes on the command line after you call your application executable. 
So ./yourapp-opt -i input_file.i --allow-test-objects

**Reply by EdSheltonUKAEA** [reply]
Thanks.  That's exactly what I had tried, but it doesn't run and gives the following error message:

> *** ERROR ***
> A 'InterfaceDiffusion' is not a registered object.
> 
> If you are trying to find this object in a dynamically linked library, make sure that
> the library can be found either in your "Problem/library_path" parameter or in the
> MOOSE_LIBRARY_PATH environment variable.

So what exactly is it I need to have in one of those "library_path" options, and where do I do this?  (Again, apologies for my ignorance in how to get the computer science aspects of MOOSE to work.)

In the meantime, I'll maybe go back to seeing if I can adapt the example file to use the InterfaceDiffusionBoundaryTerm kernel instead.

**Reply by GiudGiud** [reply]
Ah you are not using moose but your application. Ok you do not have access to moose test objects then. You would need to move InterfaceDiffusion from moose/test/(src/include) to your application to use it

**Reply by GiudGiud** [reply]
OR
use app_type = MooseTestApp and specify the library_path to the MOOSE test library

**Comment by EdSheltonUKAEA** [comment]
Hello!
I'm following up on this thread from a couple of years ago to ask further questions about the InterfaceReaction interface kernel which we're now trying to use again.

I think there might be an error with how the maths of the forwards and backwards reactions is presented on the website documentation page here: [https://mooseframework.inl.gov/source/interfacekernels/InterfaceReaction.html]

Should it be: dC_2/dt = - dC_1/dt = k_f * C_1 - k_b * C_2

As it is written (dC_1/dt = k_f * C_1 - k_b * C_2) this means that for k_b=0, then the maths says a positive value of k_f results in C_1 getting bigger and C_2 getting smaller, which is the opposite of physically what should happen.

Related to this I have looked at the source file for the kernel to see how the equation is implemented there. I'm not a C++ coder, but from what I can see, the relevant equations are implemented as shown on the documentation page:

    Real
    InterfaceReaction::computeQpResidual(Moose::DGResidualType type)
    {
    Real r = 0;
    switch (type)
    {
    // Move all the terms to the LHS to get residual, for primary domain
    // Residual = kfu - kbv = kfu - kbv
    // Weak form for primary domain is: (test, kfu - kbv)
    case Moose::Element:
    r = _test[_i][_qp] * (_kf * _u[_qp] - _kb * _neighbor_value[_qp]);
    break;

    // Similarly, weak form for secondary domain is: -(test, kf*u - kb*v),
    // flip the sign because the direction is opposite.
    case Moose::Neighbor:
      r = -_test_neighbor[_i][_qp] * (_kf * _u[_qp] - _kb * _neighbor_value[_qp]);
      break;

    }
    return r;
    }

A negative sign is included in the second half of this block of code, but this is explicity to capture that the secondary domain concentration goes the opposite direction to the primary domain concentration.

However, in spite of the "missing" negative sign for the forward reaction, when I run it in a model, the solution seems to be physically right. With k_f positive and k_b = 0, the concentration of C_1 in the primary domain side of the interface decreases and the concentration of C_2 in the secondary domain on the "neighbour" side increases.

Can anyone explain what's going on here and why it works?

**Reply by GiudGiud** [reply]
Hello

Everything gets moved to the left hand side when we implement computeQpResidual.

so this `dC1/dt = k_f * C_1 - k_b * C_2` is really `dC1/dt  -  k_f * C_1 + k_b * C_2= 0`

this causes a few sign changes depending on how people were writing their equations.
But yes here the documentation is wrong comparing to what's implemented.
The doc says

```
dC1/dt  -  k_f * C_1 + k_b * C_2= 0
```
and what's implemented (and what we actually want) is
```
dC1/dt  +  k_f * C_1 - k_b * C_2= 0
```

Thanks for reaching out! I ll put up a patch

Guillaume

**Reply by EdSheltonUKAEA** [reply]
Ah, thank you.  Suddenly it all becomes clear!
So now I know that the interface kernel with positive values for k_f and k_b works as expected, and why it works!

Incidentally, we find that running models with just InterfaceReaction seem to give sensible results, and it doesn't seem absolutely necessary to include the InterfaceDiffusion interface kernel as well.  In fact including InterfaceDiffusion can cause problems.  Is it OK to do this?

**Reply by GiudGiud** [reply]
Well what is physically happening in the system you are modeling?
Is the diffusive flux supposed to be continuous across the interface? Or is there a reaction? Or both

**Reply by lindsayad** [reply]
If your only term that is integrated by parts is a diffusion term, then this interface kernel is weakly enforcing the conditions

`- Du \nabla u * n = kf * u - kb * v = Dv * \nabla v * n`

where u is the variable on one half of the domain, v is the variable on the other half, and Du and Dv are the respective diffusion coefficients, and n is the normal vector pointing out of the u domain
